src/System.Private.Reflection.Core/src/System/Reflection/Runtime/General/Dispensers.Ecma.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using System.Collections.Generic;
+
+using System.Reflection.Runtime.General;
+using System.Reflection.Runtime.TypeInfos;
+using System.Reflection.Runtime.Assemblies;
+using System.Reflection.Runtime.Dispensers;
+using System.Reflection.Runtime.PropertyInfos;
+
+using Internal.Reflection.Core;
+using Internal.Reflection.Core.Execution;
+
+using System.Reflection.Metadata;
+
+//=================================================================================================================
+// This file collects the various chokepoints that create the various Runtime*Info objects. This allows
+// easy reviewing of the overall caching and unification policy.
+//
+// The dispenser functions are defined as static members of the associated Info class. This permits us
+// to keep the constructors private to ensure that these really are the only ways to obtain these objects.
+//=================================================================================================================
+
+namespace System.Reflection.Runtime.Assemblies
+{
+    //-----------------------------------------------------------------------------------------------------------
+    // Assemblies (maps 1-1 with a MetadataReader/ScopeDefinitionHandle.
+    //-----------------------------------------------------------------------------------------------------------
+    internal partial class RuntimeAssembly
+    {
+        static partial void GetEcmaRuntimeAssembly(AssemblyBindResult bindResult, ref RuntimeAssembly runtimeAssembly)
```

Comment:
In order to remain faithful to the dispenser.cs pattern (don't allow construction that bypasses our caching/unification logic), NativeFormatRuntimeAssembly's constructor should be private and the dispenser code for it made part of NativeFormatRuntimeAssembly's class.